### ISSUE TITLE
[chore] Clean up and remove `utils` in `linkml` and `linkml-runtime`

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#2371]
Related to: https://github.com/linkml/linkml/issues/1784

**What is your feature request?**

I am actually a defender of `utils` modules as a valid pattern, where sometimes small routines genuinely have no good place to go, but I think that the scale of `utils` modules in both `linkml` and `linkml-runtime` is pretty extreme and proving detrimental to sustainability/maintenance. 

Problems with gigantic utils modules:
- Can't find stuff when you want it
- So you reimplement it
- Over and over.
- Forever.

This has happened to me personally about a dozen times, and I think it's a pretty strong turnoff to new developers who are scoping out the project and see "whoa nelly that is a big junk drawer" - as much as i adore the somewhat 'by the seat of the pants' nature of linkml, it's a sign of bad code health, and i have had several people express to me candidly that it was enough for them to not consider linkml (not trying to be harsh, but also not kidding). It generally makes adding new features harder than it needs to be (hunting through utils to see if something has already been written) and makes maintenance challenging (something will turn out to have its root cause in a utils module that hasn't been touched in years and we frankly forgot about).

This is especially true when there are seemingly core classes and functions in `utils`. The most obvious ones to me are `Generator` (which should go in `linkml.generators`)  and `Schemaview` (which should be split off into its own package imo). 

I think there needs to be a relatively large refactoring to split up
- "core" linkml representation: all the stuff that is needed to evaluate a linkml schema and return a fully induced schema
- generators, each of which might have their own runtime package
- dumpers/loaders (or combine with generator packages)
- validation/linting
- transformation/conversion
- cli
because otherwise these things get splayed out and mashed together and we end up with a million dep package,
but to get there we need to start tidying up `utils` :)

Here's what i would do

`linkml`
- cli_utils.py -> `cli`
- converter.py -> `cli`
- datautils.py -> wherever loaders/dumpers go
- datavalidator.py -> `validator`
- deprecation.py -> either stays or in package root
- exceptions.py -> either stays or in package root
- execute_tutorial.py -> I actually am not sure because I think this should just be doctests/`myst_nb`
- generator.py -> `generators`
- helpers.py -> remove
- logictools.py -> `transformers`
- mergeutils.py -> I think this is mostly deprecated post schemaview?
- rawloader.py -> dumpers/loaders
- schema_builder.py -> I don't really understand the use of this beyond just instantiating the metamodel classes directly, but put in new module `schema` for tools for interacting with a schema itself
- schema_fixer.py -> new module `schema`
- schemaloader.py -> loaders/dumpers
- schemasynopsis.py -> new module `schema`
- sqlutils.py -> dumpers/loaders
- typereferences.py -> not sure what this is, but seems like it should go in new module `schema`
- validation.py -> `validator`

`linkml-runtime`
I continue to not be sure what the division of labor re: `linkml-runtime` is, I thought it was supposed to be accessory classes for pythongen models but there's a lot more in here. I think this should probably get clarified and a decent bulk of this should get moved to `linkml` or split out into a separate `schemaview` package so that the downstream generators can slim down their deps. 

- comparefiles.py -> new module `cli`
- compile_python.py -> not used in `linkml-runtime` as far as i can tell, maybe move to `linkml.generators.python`
- context_utils.py -> probably `linkml.generators.jsonldgen`
- csvutils.py -> linkml dumpers/loaders
- curienamespace.py -> new module `linkml_runtime.types` to contain types used in generated pythongen modules
- dictutils.py -> rm
- distroutils.py -> rm
- enumerations.py -> `linkml_runtime.types.enum`
- eval_utils.py -> not sure what this is, but it looks like what i might have been looking for awhile! -> `linkml.schema`
- formatutils.py -> some of these go to `linkml.generators.common.naming`, others (eg. remove empty, to dict) should go to `linkml_runtime.methods` and get attached to generated pythongen models via `YAMLRoot`
- inference_utils.py -> not used in `linkml-runtime` as far as i can tell, so `linkml.inference`
- introspection.py -> rm (does this work?)
- metamodelcore.py -> `linkml_runtime.types`
- namespaces.py -> package root (`linkml_runtime.namespaces`)
- pattern.py -> Not sure what this does!
- permissiblevalueimpl.py -> `linkml_runtime.types.enum`
- ruleutils.py -> `linkml.schema`
- schema_as_dict.py -> rm
- schema_builder.py -> oh dang is this a fully duplicated module? probably goes in `linkml` right?
- schemaops.py -> schemaview
- schemaview.py -> own package `linkml-schema` (that arguably all the `linkml.schema` stuff should go in)
- schemaview_cli.py -> `linkml-schema.cli`
- slot.py -> `linkml_runtime.types`
- strictness.py -> rm
- uri_validator.py -> holy crap this is awesome and it should be its own package
- walker_utils.py -> not sure, this one seems like genuine `utils` material
- yamlutils.py -> `YAMLRoot` should be renamed something like `PythonGenRoot` and moved to `linkml_runtime.types.root`, stuff that's specifically for dealing with pythongen classes stays with it, other stuff goes in dumpers/loaders.

Basically what i'm saying is that i would do this if i had some assurance that the big gigantic PR would be merged. 

To ease transition what i would do is make a transitional `MetaPathFinder` with a dict mapping old import locations to the new locations that emits a deprecation warning using the `deprecation` module so we don't need to leave a ton of empty skeleton modules lying around. 

This would involve creating at least one new package (ideally `linkml-schema` which contains schemaview et al minimal tooling required to materialize a schema) but the whole point of this project is to trim dead code and decrease maintenance burden - eg. rather than relying on taking up all of Chris's precious time, we could split off different generators/etc. as being taken care of by other people, and isolate core linkml changes to the `linkml-schema` module. We could keep all the same testing rigorousness and test each package along with all the others the same way we do with `linkml-runtime`, and we could even co-version them together if we wanted to.

Anyway, i've been sorta wanting to do a major cleanup for awhile and wanted to start sketching a plan, because otherwise i feel like we're sorta at a point where things could inflect off into heat death since i've seen an increasing number of reimplementing basic functions in-place because the canonical impl can't be found.

**How important is this feature to you?** Select from the options below:
• Medium - can do work without it; but it's important (e.g. to save time or for convenience)

**Comment by sneakers-the-rat** [comment]
Gonna keep a running list of stuff that appears to have no usage:

- `linkml_runtime.index` - getting a bunch of type errors and linting errors when i'm trying to get `linkml_runtime` up to `linkml` ruff rules.

**Comment by sierra-moxon** [comment]
from dev call:
- definitely need to get rid of the cruft, definitely need to minimize the utils scripts
- some hesitation to do this kind of a big refactor, but open to it - def needs to be coordinated with the team before we spend a lot of time on a PR that needs concentrated effort to review/etc...

**Comment by cmungall** [comment]
Minor:

`linkml_runtime.index` was flagged as potentially unused. This is being used in `linkml-map`, and is potentially very useful much more generally. The fact that it is tucked away in this was does lend evidence to the need for the refactor...

**Comment by sneakers-the-rat** [comment]
> def needs to be coordinated with the team before we spend a lot of time on a PR that needs concentrated effort to review/etc...

Yes! This is what i was thinking. that we could use this issue (or elsewhere!) to make a plan about what goes where, and then we can piece away at it like we did/are doing with converting unittest -> pytest. 

>  This is being used in `linkml-map`, and is potentially very useful much more generally. 

Yes! Also part of the idea is to highlight all the cool things that are tucked away in `utils` and make them discoverable/clear how and when to use them. Along with this we can do a bit of a refactor of the docs and make a clear api docs section, where that is currently mixed in with the Generators section, the CLI section, the Developers Guide, and the Code section.

**Comment by Silvanoc** [comment]
Is something happening around the topics mentioned in this PR? I see a real need to address this topic, but no movement for the last 2 months...

**Comment by Silvanoc** [comment]
> basically i'm just waiting on any input on the above sheet - we can start picking it off one by one in separate PRs. maybe an easy way to start would be for core maintainers to pick the easy/uncontroversial ones and say "these ones yes" so we can get started on those while the ones that might need some discussion can have it

Well, then [this question](https://github.com/linkml/linkml/issues/2382#issuecomment-2621122799) also applies here:

> who is a "core maintainer"? Are they listed somewhere? Apart from @cmungall I could not name any...

I have the feeling that you @sneakers-the-rat are having a similar issue to me. You are trying to appeal someone (a "core maintainer") that is not being appealed by your messages. They are probably not reading all comments (they would get nuts) and not knowing them (by name) they can not be mentioned to ensure that they are appealed.

**Comment by sneakers-the-rat** [comment]
ha ya, i usually take the people whose [names are on the tin](https://github.com/linkml/linkml/blob/8938fb1f93737d0e3182b3640e208d83f2e1e1fb/pyproject.toml#L5-L15) as the people who can make calls about a package, but i think linkml is a little freeform in that regard (and yes ppl are busy)

**Comment by Silvanoc** [comment]
@sneakers-the-rat the symbols that you've used in the "status" column of your table are not rendered in none of my browsers (Firefox, Safari, Chrome), are they so uncommon?

I'd propose at least one change in the table: `cli_utils.py` does not go to `linkml.cli` (which is an umbrella importing all other CLIs), but to `linkml.generators.generator` (gets imported by all generators, since provide the base classes).

**Comment by sneakers-the-rat** [comment]
> the symbols that you've used in the "status" column of your table are not rendered in none of my browsers (Firefox, Safari, Chrome), are they so uncommon?

they are likely rendering correctly, it is literally [`U+2370` ](https://codepoints.net/U+2370?lang=en) bc i don't know whether any of them are approved or not lol

> you do not propose linkml/cli/main.py, but rather a submodule of linkml.cli, right?

yeah, somewhere in there.

**Comment by Silvanoc** [comment]
> uri_validator.py 	⍰ 	(holy crap this is awesome and it should be its own package)

@sneakers-the-rat thanks for the appraisal, I'm the author 😄 

IMO it could even have a live outside of LinkML, but it would require more effort than what I can invest on it.

**Comment by sneakers-the-rat** [comment]
> do we really need a file comparison command in the runtime?!

i agree. i personally would rm

**Comment by Silvanoc** [comment]
> > do we really need a file comparison command in the runtime?!
> 
> i agree. i personally would rm

@sneakers-the-rat would you then please update the target column into "rm"?

**Comment by Silvanoc** [comment]
I fully agree with the proposal of moving `slot`, `metamodelcore`,... from `utils` to `types`, and therefore started https://github.com/linkml/linkml-runtime/pull/396. I see a major issue and blocker:

Getting the changes that this PR tries to address effectively working requires a synchronized effort involving the repos `linkml`, `linkml-runtime` and `linkml-model`. Because the Python generator needs to be adapted so that generated classes take the new path and the model needs to be also generated with the new paths. It's questionable if the effort is worth the gain...

**Comment by Silvanoc** [comment]
@sneakers-the-rat I've edited the table in your issue description to update the current status. I hope it's fine for you.

**Comment by matentzn** [comment]
@sneakers-the-rat I assigned you to this issue, because it seems to me you are in the drivers seat - let me know if thats not the case and I will assign to someone else. Looking at open PRs next.
